### ISSUE TITLE
Don't crash on attepmt to access an empty diagnostics area

### ISSUE THREAD

**Issue by Totktonada** [tarantool/tarantool#4737]
Tarantool version: 2.4.0-13-ga16637365.

I propose to add a 'lack of an error' error and return it from `diag_get()` instead of NULL. This error object can be created once and saved in a static variable. `diag_is_empty()` can compare a current error with this 'lack of an error' error.

When `diag_get()` is called for an empty diagnostics area, this is surely mistake in our code. It worth to log the problem (maybe even log a backtrace) and continue. A user may get 'lack of an error' diagnostics in the case, but it is surely better than crash on NULL dereference.

The idea appears during investigation of #4730.

**Comment by Totktonada** [comment]
BTW, if we'll create a test case that will check that #4730 does not crash tarantool, we should change it within this task to verify a log file against the 'lack of an error' message. Maybe there are other tests of this kind: we need to look through tests.

Alternative: left old behaviour (crash) for the Debug build. This however undesired, because we sometimes verify hypothesis on a client side using the Debug build, when it is hard to reproduce an issue on our side.

**Comment by Totktonada** [comment]
(A bit more thoughts.)

My proposal is based on the following point: when a diagnostics is not set (or was unexpectedly cleared with diag_clear() / diag_move()), a flow of the code still right, because it depends on return code rather than the diagnostic itself. So tarantool able to handle this situation gracefully and report an error w/o a specific message: just report that an error occurs during execution of a certain function (that is why I proposed to show backtrace).

This point however does not work when a control flow depends of a last error existence (`diag_is_empty()`) or type. We should look into such places very carefully if we'll move this way: how the control flow may be altered by a diag existence / type, is taking of a wrong branch may lead to data corruption? If there is such possibility, then we should fail.

I don't so sure now that the proposal is right. I need to look into the code deeper.

There were discussions around this proposal within the team and I want to answer one objection. This sounds like so: if tarantool does not crash in the case of missed diag_set(), then we'll not receive a report about the problem from a user. It is so if the diagnostics is not important for the user (there are reported cases when they are important, say, #4138). If it is not important for the user, then it is not something that deserves a service outage or degradation due to the crash. We saw it already: #4730 makes tarantool unusable, because of very often crashes.

It worth also to note that logging of such errors still allow us to catch them with our tests.

**Comment by kostja** [comment]
Of course, put some lipstick on this pig. Of course @kyukhin is in favor. Apparently it's some kind of Intel way (oops what happened to them...)

**Comment by kostja** [comment]
Once Tarantool stopped using exceptions, this kind of issue became inevitable. The proper way to handle it is to panic and crash, not try to be nice. It is possible to cover up the crash in the release build with a backtrace in the log and a ENOERR error value, set on access. What you suggest - always having this default is just a signature of being unable to trace down the original issue and not wanting to either.

**Comment by cyrillos** [comment]
Kostya, it is unclear from text above but we definitely won't be hiding problems. Initially we thought of more grace exit instead of plain SIGSEGV, but exit, not continue.

**Comment by Totktonada** [comment]
It is not unclear: I proposed to don't exit when control flow does not depend of a kind of an error (or its absence) and we detected lack of a diagnostics, but we know what to do next. Don't sure however whether it is possible to verify the whole code for branching based on an error type / existence.

Maybe the amount of work is on the order of check the whole code against proper manipulations with diagnostics area and so is not worth to do.
